packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart

```diff
@@ -0,0 +1,149 @@
+  info,
+
+  /// Logs fine logs at the most.
+  ///
+  /// A fine message is one that is not important for logging outside of
+  /// debugging potential issues in the application.
+  fine,
+
+  /// Logs everything.
+  all,
+}
+
+/// Signature of a function that logs a [LogMessage].
+typedef void LoggingFunction(LogMessage log);
+
+/// The default logging function.
+///
+/// Runs the [print] function using the format string:
+///   '[${log.levelName}]::${log.tag}--${log.time}: ${log.message}'
+///
+/// Exits with status code 1 if the `log` is [LoggingLevel.severe].
+LoggingFunction defaultLoggingFunction = (LogMessage log) {
+  print('[${log.levelName}]::${log.tag}--${log.time}: ${log.message}');
+  if (log.level == LoggingLevel.severe) {
+    exit(1);
+  }
+};
+
+/// Represents a logging message created by the logger.
+///
+/// Includes a message, the time the message was created, the level of the log
+/// as an enum, the name of the level as a string, and a tag. This class is used
+/// to print from the global logging function defined in
+/// [Logger.loggingFunction] (a function that can be user-defined).
+class LogMessage {
+  /// The actual log message.
+  final String message;
+
+  /// The time the log message was created.
+  final DateTime time;
+
+  /// The level of this log.
+  final LoggingLevel level;
+
+  /// The human readable level of this log.
+  final String levelName;
+
+  /// The tag associated with the message. This is set to [Logger.tag] when
+  /// emitted by a [Logger] object.
+  final String tag;
+
+  /// Creates a log, including the level of the log, the time it was created,
+  /// and the actual log message.
+  ///
+  /// When this message is created, it sets its [time] to [DateTime.now].
+  LogMessage(this.message, this.tag, this.level)
+      : this.levelName =
+            level.toString().substring(level.toString().indexOf('.') + 1),
+        this.time = new DateTime.now();
+}
+
+/// Logs messages using the global [LoggingFunction] and logging level.
+///
+/// Example of setting log level to [LoggingLevel.warning] and creating a
+/// logging function:
+///
+/// Logger.globalLevel = LoggingLevel.warning;
+class Logger {
+  /// The tag associated with the log message (usable in the logging function).
+  /// [LogMessage] objects emitted by this class will have [LogMessage.tag] set
+  /// to this value.
+  final String tag;
+
+  /// Determines what to do when the [Logger] creates and attempts to log a
+  /// [LogMessage] object.
+  ///
+  /// This function can be reassigned to whatever functionality of your
+  /// choosing, so long as it has the same signature of [LoggingFunction] (it
+  /// can also be an asynchronous function, if doing file I/O, for
+  /// example).
+  static LoggingFunction loggingFunction = defaultLoggingFunction;
+
+  /// Determines the logging level all [Logger] instances use.
+  static LoggingLevel globalLevel = LoggingLevel.none;
+
+  /// Creates a logger with the given [tag].
+  Logger(this.tag);
```

Comment:
constructors should be first, so that we can quickly tell if there's an implicit no-argument constructor without reading the whole definition of the class

Reply:
Fixed.